### PR TITLE
✨ Computational backend: set container limits as labels ⚠️ (devops checks on grafana dashboards!)

### PR DESCRIPTION
**Pull request by sanderegg**
## What do these changes do?
- [x] add container cpu limits and memory limits on container (both for dynamic-sidecar services and computational services)
- [x] uniformize labels for dynamic-sidecar services + computational services
- [x] all labels used now are prefixed with *io.simcore.runtime.* (as in *io.simcore.runtime.node-id* or *io.simcore.runtime.project-id*) to properly differentiate from simcore image labels which are static
- [x] backwards compatible with old nomenclature (which should be removed once this PR reaches production + 1)
- [x] simplified dynamic-sidecar labels by cleaning up unused or misused ones
- ⚠️: possibly some grafana dashboards might need to be adapted. (i.e. *simcore_user_agent* changed to *io.simcore.runtime.simcore-user-agent*)

**NOTE that underscores are not considered as valid docker label keys!**

### legacy services now:
image

### computational service now:
image

### dynamic-services now:
show the same labels in service and container labels.
NOTE: io.simcore.scheduler-data is still the same. I did not dare touch this now.
NOTE2: docker network/volumes labels are not changed. I will let @GitHK take care of this or when I get again an incentive to touch this.

## Related issue/s
- resolves https://github.com/ITISFoundation/osparc-simcore/issues/3638

### CODE REVIEW

File: services/director-v2/tests/unit/with_dbs/test_modules_dynamic_sidecar_docker_service_specs.py [hunk 1]
```diff
@@ -132,10 +133,10 @@ def expected_dynamic_sidecar_spec(
...
-                        "outputs_path": "/tmp/outputs",
-                        "state_exclude": ["/tmp/strip_me/*", "*.py"],
-                        "state_paths": ["/tmp/save_1", "/tmp_save_2"],
+                        "inputs_path": "/tmp/inputs",  # noqa: S108
+                        "outputs_path": "/tmp/outputs",  # noqa: S108
+                        "state_exclude": ["/tmp/strip_me/*", "*.py"],  # noqa: S108
+                        "state_paths": ["/tmp/save_1", "/tmp_save_2"],  # noqa: S108
                     },
                     "product_name": osparc_product_name,
                     "project_id": "dd1d04d9-d704-4f7e-8f0f-1ca60cc771fe",
```
Comment (mrnicegyu11) [inline services/director-v2/tests/unit/with_dbs/test_modules_dynamic_sidecar_docker_service_specs.py:142]: I guess in this whole section one could make heavy use of `faker`, maybe for another time

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_service_specs/sidecar.py [hunk 2]
```diff
@@ -268,12 +278,16 @@ def get_dynamic_sidecar_spec(
...
-                    "uuid": f"{scheduler_data.node_uuid}",
-                },
+                "Labels": StandardSimcoreDockerLabels(
+                    user_id=scheduler_data.user_id,
+                    project_id=scheduler_data.project_id,
+                    node_id=scheduler_data.node_uuid,
+                    product_name=scheduler_data.product_name,
+                    simcore_user_agent=scheduler_data.request_simcore_user_agent,
+                    swarm_stack_name=dynamic_sidecar_settings.SWARM_STACK_NAME,
+                    memory_limit=ByteSize(0),  # this should get overwritten
```
Comment (pcrespov) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_service_specs/sidecar.py:288]: memory and cpu limits are set here to zero which is what you called UNDEFINED above, right? 
So this undefined fits well to set "limits" values but not somehting else

File: services/director/src/simcore_service_director/producer.py [hunk 3]
```diff
@@ -129,6 +131,13 @@ async def _read_service_settings(
     return settings
 
 
+_SIMCORE_RUNTIME_DOCKER_LABEL_PREFIX: str = "io.simcore.runtime."
+
+
+def _to_simcore_runtime_docker_label_key(key: str) -> str:
+    return f"{_SIMCORE_RUNTIME_DOCKER_LABEL_PREFIX}{key.replace('_', '-').lower()}"
```
Comment (mrnicegyu11) [inline services/director/src/simcore_service_director/producer.py:138]: See above, this is good but knowledge about why we replace might fade if there is no comment

**Review by pcrespov — APPROVED** [review]
:+1:

**Review by mrnicegyu11 — APPROVED** [review]
very promissing, and very good for cleanup pruposes on the labels! thanks a lot!

I have only added minors and some questions for my learndings.

Please add a big verbose devops warning to the title of the PR etc. , for sure we have to modify some PromQl in the grafan dashboards, and I will prepare an ops-commit once this is in master (and I see what breaks) that shall propagate with this PR

File: packages/models-library/src/models_library/service_settings_labels.py [hunk 4]
```diff
@@ -201,27 +202,25 @@ def validate_volume_limits(cls, v, values) -> str | None:
...
             outputs_path: Path | None = values.get("outputs_path")
             state_paths: list[Path] | None = values.get("state_paths")
             path = Path(path_str)
             if not (
-                path == inputs_path
-                or path == outputs_path
-                or (state_paths is not None and path in state_paths)
+                path in (inputs_path, outputs_path)
+                or state_paths is not None
+                and path in state_paths
```
Comment (GitHK) [inline packages/models-library/src/models_library/service_settings_labels.py:215]: are you sure this is equivalent? I think it is not. I think it should be like the following.
```suggestion
                path in (inputs_path, outputs_path)
                or (state_paths is not None
                and path in state_paths)
```

File: packages/models-library/src/models_library/docker.py [hunk 5]
```diff
@@ -21,29 +31,176 @@ class DockerGenericTag(ConstrainedStr):
...
+    "project_id": f"{_SIMCORE_RUNTIME_DOCKER_LABEL_PREFIX}project-id",
+    "simcore_user_agent": f"{_SIMCORE_RUNTIME_DOCKER_LABEL_PREFIX}simcore-user-agent",
+    "study_id": f"{_SIMCORE_RUNTIME_DOCKER_LABEL_PREFIX}project-id",
+    "user_id": f"{_SIMCORE_RUNTIME_DOCKER_LABEL_PREFIX}user-id",
+    "uuid": f"{_SIMCORE_RUNTIME_DOCKER_LABEL_PREFIX}node-id",
+    "mem_limit": f"{_SIMCORE_RUNTIME_DOCKER_LABEL_PREFIX}memory-limit",
+    "swarm_stack_name": f"{_SIMCORE_RUNTIME_DOCKER_LABEL_PREFIX}swarm-stack-name",
+}
+_UNDEFINED_LABEL_VALUE_STR: Final[str] = "undefined"
+_UNDEFINED_LABEL_VALUE_INT: Final[str] = "0"
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply packages/models-library/src/models_library/docker.py:47]: this is only for backwards compatibility and will disappear soon after. I'd rather not spend too much time on this.

File: packages/models-library/src/models_library/service_settings_labels.py [hunk 6]
```diff
@@ -138,8 +139,8 @@ class Config(_BaseConfig):
                     "value": [
                         {
                             "ReadOnly": True,
-                            "Source": "/tmp/.X11-unix",  # nosec
-                            "Target": "/tmp/.X11-unix",  # nosec
+                            "Source": "/tmp/.X11-unix",  # nosec  # noqa: S108
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply packages/models-library/src/models_library/service_settings_labels.py:142]: yes, the ruff linter is not compatible with ```# nosec``` see [https://github.com/tylerwince/flake8-bandit/issues/20](https://github.com/tylerwince/flake8-bandit/issues/20) as it implements flake8-bandit and not bandit itself.

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api/_core.py [hunk 7]
```diff
@@ -222,58 +203,90 @@ async def get_dynamic_sidecar_state(service_id: str) -> tuple[ServiceState, str]
...
 
+async def _list_docker_services(
+    client: aiodocker.docker.Docker,
+    *,
+    node_id: NodeID | None,
+    swarm_stack_name: str,
+    return_only_sidecars: bool,
+) -> list[Mapping]:
+    # NOTE: this is here for backward compatibility when first deploying this change.
+    # shall be removed after 1-2 releases without issues
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api/_core.py:250]: yes that is why I added the warning below.

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_service_specs/settings.py [hunk 8]
```diff
@@ -148,11 +155,28 @@ def update_service_params_from_settings(
                 ].extend(mount_settings)
 
     container_spec = create_service_params["task_template"]["ContainerSpec"]
-    # set labels for CPU and Memory limits
-    container_spec["Labels"]["nano_cpus_limit"] = str(
-        create_service_params["task_template"]["Resources"]["Limits"]["NanoCPUs"]
+    # set labels for CPU and Memory limits, for both service and container labels
+    # NOTE: cpu-limit is a float not NanoCPUs!!
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_service_specs/settings.py:159]: hmm not sure what you mean here @mrnicegyu11?
CPU is here a nanocpu, i.e. 2 CPUs = 2 * 1*10**9 and is an int. the label shows a float and the equivalent is then 2.0 CPUs

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_service_specs/sidecar.py (hunk 2 above)
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_service_specs/sidecar.py:288]: well I did not change the logic in here right. the point here is to standardize the labels and make them docker compatible.

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_core/_events.py [hunk 9]
```diff
@@ -443,14 +442,13 @@ async def action(cls, app: FastAPI, scheduler_data: SchedulerData) -> None:
...
-            raise ValueError(
-                "Did not expect None for any of the following: "
-                f"{scheduler_data.dynamic_sidecar.dynamic_sidecar_id=} "
-                f"{scheduler_data.dynamic_sidecar.dynamic_sidecar_network_id=} "
-                f"{scheduler_data.dynamic_sidecar.swarm_network_id=} "
-                f"{scheduler_data.dynamic_sidecar.swarm_network_name=} "
-                f"{scheduler_data.proxy_admin_api_port=}"
+            msg = (
+                f"Did not expect None for any of the following:"
+                f" scheduler_data.dynamic_sidecar.dynamic_sidecar_id={scheduler_data.dynamic_sidecar.dynamic_sidecar_id!r} "
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_core/_events.py:447]: well to be honnest this was an autofix from Ruff.
but if you check the docs on f-string, !r uses the *repr* of the value given. it would actually be more explicit to just write *repr(value)* instead of *!r*.

File: services/director-v2/tests/unit/test_modules_dask_client.py [hunk 10]
```diff
@@ -471,7 +472,14 @@ def comp_run_metadata(faker: Faker) -> MetadataDict:
 
 @pytest.fixture
 def task_labels(comp_run_metadata: MetadataDict) -> ContainerLabelsDict:
-    return comp_run_metadata
+    return parse_obj_as(
+        ContainerLabelsDict,
+        {
+            k.replace("_", "-").lower(): v
```
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/director-v2/tests/unit/test_modules_dask_client.py:478]: So, I think the comment is not needed because:
- *parse_obj_as* is a pydantic function that parses the value into a ContainerLabelsDict, which is a dict[DockerLabelKey, str]
- *DockerLabelKey* is a pydantic *ConstrainedStr* which is a str that uses a regex to impose the Docker label key rules (i.e. no *_* but only *-*)
Is this ok?

File: services/director-v2/tests/unit/with_dbs/test_modules_dynamic_sidecar_docker_service_specs.py (hunk 1 above)
Comment (sanderegg) [reply services/director-v2/tests/unit/with_dbs/test_modules_dynamic_sidecar_docker_service_specs.py:142]: I agree to both.

File: services/director/src/simcore_service_director/producer.py (hunk 3 above)
Comment (sanderegg) [reply services/director/src/simcore_service_director/producer.py:138]: ok. this one is special. that is the director-v0 which is scheduled to be retired. I do not want to invest to much time there, I just made it compatible. Otherwise I would agree.

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api/_core.py [hunk 11]
```diff
@@ -74,7 +69,7 @@ async def create_network(network_config: dict[str, Any]) -> NetworkId:
             network_name = network_config["Name"]
             # make sure the current error being trapped is network dose not exit
             if f"network with name {network_name} already exists" not in str(e):
-                raise e
+                raise
```
Comment (GitHK) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api/_core.py:72]: Question: why this change? is it not better to `raise e`?

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api/_core.py [hunk 12]
```diff
@@ -201,17 +184,15 @@ async def _get_task_data_when_service_running(service_id: str) -> Mapping[str, A
         service_state = task["Status"]["State"]
 
         if service_state not in TASK_STATES_RUNNING:
-            raise TryAgain()
+            raise TryAgain
```
Comment (GitHK) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api/_core.py:187]: I think I've already expressed my disagreement on this one. Why mix classes and objects. It makes more senso to raise instances of errors and not classes. At least it is consistent.

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api/_core.py [hunk 13]
```diff
@@ -476,8 +465,8 @@ async def _update_service_spec(
                         e.status == status.HTTP_500_INTERNAL_SERVER_ERROR
                         and "out of sequence" in e.message
                     ):
-                        raise TryAgain() from e
-                    raise e
+                        raise TryAgain from e
```
Comment (GitHK) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api/_core.py:468]: again here using class instead of an instance.

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_service_specs/proxy.py [hunk 14]
```diff
@@ -79,12 +80,18 @@ def get_dynamic_proxy_spec(
...
+        }
+        | StandardSimcoreDockerLabels(
+            user_id=scheduler_data.user_id,
+            project_id=scheduler_data.project_id,
+            node_id=scheduler_data.node_uuid,
+            product_name=scheduler_data.product_name,
+            simcore_user_agent=scheduler_data.request_simcore_user_agent,
+            swarm_stack_name=dynamic_sidecar_settings.SWARM_STACK_NAME,
+            memory_limit=ByteSize(MEMORY_50MB),
+            cpu_limit=float(CPU_10_PERCENT) / (1.0 * 10**9),
```
Comment (GitHK) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_service_specs/proxy.py:93]: May I suggest this one?
```suggestion
            cpu_limit=CPU_10_PERCENT / 1e9,
```

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_service_specs/proxy.py [hunk 15]
```diff
@@ -93,12 +100,16 @@ def get_dynamic_proxy_spec(
...
-                },
+                "Labels": StandardSimcoreDockerLabels(
+                    user_id=scheduler_data.user_id,
+                    project_id=scheduler_data.project_id,
+                    node_id=scheduler_data.node_uuid,
+                    product_name=scheduler_data.product_name,
+                    simcore_user_agent=scheduler_data.request_simcore_user_agent,
+                    swarm_stack_name=dynamic_sidecar_settings.SWARM_STACK_NAME,
+                    memory_limit=ByteSize(MEMORY_50MB),
+                    cpu_limit=float(CPU_10_PERCENT) / (1.0 * 10**9),
```
Comment (GitHK) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_service_specs/proxy.py:111]: same suggestion as above

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_service_specs/settings.py [hunk 16]
```diff
@@ -105,22 +128,6 @@ def update_service_params_from_settings(
...
-        elif param.setting_type == "EndpointSpec":
-            if "Ports" in param.value:
-                if (
-                    isinstance(param.value["Ports"], list)
-                    and "TargetPort" in param.value["Ports"][0]
-                ):
-                    create_service_params["labels"]["port"] = create_service_params[
-                        "labels"
-                    ]["service_port"] = str(param.value["Ports"][0]["TargetPort"])
-
```
Comment (GitHK) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_service_specs/settings.py:123]: Question: are these no longer useful? This was only used to extract the port aftewards?

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_core/_events.py [hunk 17]
```diff
@@ -443,14 +442,13 @@ async def action(cls, app: FastAPI, scheduler_data: SchedulerData) -> None:
...
             or scheduler_data.proxy_admin_api_port is None
         ):
-            raise ValueError(
-                "Did not expect None for any of the following: "
-                f"{scheduler_data.dynamic_sidecar.dynamic_sidecar_id=} "
-                f"{scheduler_data.dynamic_sidecar.dynamic_sidecar_network_id=} "
-                f"{scheduler_data.dynamic_sidecar.swarm_network_id=} "
-                f"{scheduler_data.dynamic_sidecar.swarm_network_name=} "
-                f"{scheduler_data.proxy_admin_api_port=}"
+            msg = (
```
Comment (GitHK) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_core/_events.py:445]: I'm not sure why, but this is less readable now.

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_core/_events.py [hunk 18]
```diff
@@ -489,6 +487,7 @@ async def action(cls, app: FastAPI, scheduler_data: SchedulerData) -> None:
             project_id=scheduler_data.project_id,
             node_id=scheduler_data.node_uuid,
             simcore_user_agent=scheduler_data.request_simcore_user_agent,
+            swarm_stack_name=dynamic_sidecar_settings.SWARM_STACK_NAME,
```
Comment (GitHK) [inline services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_core/_events.py:490]: Are you sure this is required? Or do we ant to add this to the labels of the user services as well?

File: services/director-v2/tests/unit/with_dbs/test_modules_dynamic_sidecar_docker_api.py [hunk 19]
```diff
@@ -656,25 +635,6 @@ async def test_remove_dynamic_sidecar_network_fails(
...
-):
-    # start 2 fake services to emulate the dynamic-sidecar stack
-    for dynamic_sidecar_stack in dynamic_sidecar_stack_specs:
-        service_id = await docker_api.create_service_and_get_id(dynamic_sidecar_stack)
-        assert service_id
-
-    services = await docker_api.list_dynamic_sidecar_services(
-        dynamic_sidecar_settings, user_id=user_id, project_id=project_id
-    )
-    assert len(services) == 1
```
Comment (GitHK) [inline services/director-v2/tests/unit/with_dbs/test_modules_dynamic_sidecar_docker_api.py:675]: why was this one removed?

File: services/director-v2/tests/unit/with_dbs/test_modules_dynamic_sidecar_docker_api.py [hunk 20]
```diff
@@ -448,24 +445,6 @@ async def test_create_service(
...
-
-    service_inspect = await docker_api.inspect_service(service_id)
-
-    assert service_inspect["Spec"]["Labels"] == service_spec["labels"]
-    assert service_inspect["Spec"]["Name"] == service_spec["name"]
-    assert (
-        service_inspect["Spec"]["TaskTemplate"]["ContainerSpec"]["Image"]
-        == service_spec["task_template"]["ContainerSpec"]["Image"]
-    )
-
```
Comment (GitHK) [inline services/director-v2/tests/unit/with_dbs/test_modules_dynamic_sidecar_docker_api.py:467]: why was this removed?

File: services/director/src/simcore_service_director/producer.py [hunk 21]
```diff
@@ -357,8 +395,16 @@ async def _create_docker_service_params(
     mem_limit = str(
         docker_params["task_template"]["Resources"]["Limits"]["MemoryBytes"]
     )
-    container_spec["Labels"]["nano_cpus_limit"] = nano_cpus_limit
-    container_spec["Labels"]["mem_limit"] = mem_limit
+    docker_params["labels"][
+        _to_simcore_runtime_docker_label_key("cpu_limit")
+    ] = container_spec["Labels"][
+        _to_simcore_runtime_docker_label_key("cpu_limit")
+    ] = f"{float(nano_cpus_limit) / (1 * 10**9)}"
```
Comment (GitHK) [inline services/director/src/simcore_service_director/producer.py:402]:
```suggestion
    ] = f"{nano_cpus_limit / 1e9}"
```

File: services/director-v2/tests/unit/with_dbs/test_modules_dynamic_sidecar_docker_api.py (hunk 19 above)
Comment (GitHK) [reply services/director-v2/tests/unit/with_dbs/test_modules_dynamic_sidecar_docker_api.py:675]: This test from my point of view still has to pass.

**Review by GitHK — COMMENTED** [review]
So it's looking good. I have expressed some concerns below.
Also, did you test that service composed of multiple containers like s4l work as expected?

File: packages/models-library/src/models_library/service_settings_labels.py (hunk 4 above)
Comment (sanderegg) [reply packages/models-library/src/models_library/service_settings_labels.py:215]: do you have some tests to prove this? this was auto-fixed via Ruff. I can add the parenthesis for additional security.

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api/_core.py (hunk 11 above)
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api/_core.py:72]: see [https://beta.ruff.rs/docs/rules/verbose-raise/](https://beta.ruff.rs/docs/rules/verbose-raise/)

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api/_core.py (hunk 13 above)
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api/_core.py:468]: see [https://beta.ruff.rs/docs/rules/unnecessary-paren-on-raise-exception/](https://beta.ruff.rs/docs/rules/unnecessary-paren-on-raise-exception/) although I do not really care. I have to say I find it more concise, what is sure is that it is for sure not wrong.

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api/_core.py (hunk 12 above)
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_api/_core.py:187]: see [https://beta.ruff.rs/docs/rules/unnecessary-paren-on-raise-exception/](https://beta.ruff.rs/docs/rules/unnecessary-paren-on-raise-exception/)
I disagree with you, but since it is your code I will revert it.

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_core/_events.py (hunk 18 above)
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_core/_events.py:490]: I want to have the same everywhere, these are standard labels.

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_core/_events.py (hunk 17 above)
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_core/_events.py:445]: actually I do agree. will change it. nevertheless the comment is a bit weird...

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_service_specs/settings.py (hunk 16 above)
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/docker_service_specs/settings.py:123]: yes pretty weird he?

File: services/director-v2/tests/unit/with_dbs/test_modules_dynamic_sidecar_docker_api.py (hunk 19 above)
Comment (sanderegg) [reply services/director-v2/tests/unit/with_dbs/test_modules_dynamic_sidecar_docker_api.py:675]: useless code. I removed the list_dynamic_sidecar_services method

File: services/director-v2/tests/unit/with_dbs/test_modules_dynamic_sidecar_docker_api.py (hunk 20 above)
Comment (sanderegg) [reply services/director-v2/tests/unit/with_dbs/test_modules_dynamic_sidecar_docker_api.py:467]: useless code, I removed the inspect_service code

**Comment by sanderegg** [comment]
> So it's looking good. I have expressed some concerns below. Also, did you test that service composed of multiple containers like s4l work as expected?

I tried iSeg

File: services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_core/_events.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (sanderegg) [reply services/director-v2/src/simcore_service_director_v2/modules/dynamic_sidecar/scheduler/_core/_events.py]: changed back to original more explicit and concise version
